tests/integration-tests/tests/schedulers/test_custom_munge_key.py

```diff
@@ -0,0 +1,141 @@
+
+import pytest
+from assertpy import assert_that
+from constants import ENCODE_CUSTOM_MUNGE_KEY
+from remote_command_executor import RemoteCommandExecutor
+from utils import wait_for_computefleet_changed
+
+
+@pytest.mark.usefixtures("instance", "os")
+@pytest.mark.parametrize("use_login_node", [True, False])
+def test_custom_munge_key(
+    region,
+    pcluster_config_reader,
+    clusters_factory,
+    scheduler_commands_factory,
+    store_secret_in_secret_manager,
+    use_login_node,
+):
+    """
+    Test custom munge key config, rotate, update and remove.
+
+    This test contains two case: with LoginNodes and without LoginNodes section.
+    """
+    custom_munge_key_arn = store_secret_in_secret_manager(
+        region,
+        secret_string=ENCODE_CUSTOM_MUNGE_KEY,
+    )
+    cluster_config = pcluster_config_reader(use_login_node=use_login_node, custom_munge_key_arn=custom_munge_key_arn)
+    cluster = clusters_factory(cluster_config, upper_case_cluster_name=True)
+    remote_command_executor = RemoteCommandExecutor(cluster)
+    _test_custom_munge_key_fetch_and_decode(remote_command_executor)
+    _test_munge_key_shared(remote_command_executor)
+
+    # Test if compute node can run jobs, which means its munge key fetched successfully.
+    scheduler_commands = scheduler_commands_factory(remote_command_executor)
+    scheduler_commands.submit_command_and_assert_job_accepted(
+        submit_command_args={
+            "command": "srun sleep 1",
+            "nodes": 2,
+        }
+    )
+    # Test error message when both compute and login nodes are not stopped.
+    _test_update_munge_key_without_stop_login_or_compute(remote_command_executor)
+
+    if use_login_node:
+        remote_command_executor_login = RemoteCommandExecutor(cluster, use_login_node=True)
+        _test_custom_munge_key_fetch_and_decode(remote_command_executor_login)
+        remote_command_executor_login.close_connection()
+
+    # Stop compute fleets
+    cluster.stop()
+    wait_for_computefleet_changed(cluster, "STOPPED")
+
+    if use_login_node:
+        # Test error message when login nodes are not stopped.
+        _test_update_munge_key_without_stop_login_or_compute(remote_command_executor, compute_stopped=True)
+
+        # Update cluster with pcluster.stop_login.config.yaml to stop login nodes.
+        update_cluster_stop_login_config = pcluster_config_reader(
+            config_file="pcluster.stop_login.config.yaml",
+            custom_munge_key_arn=custom_munge_key_arn,
+        )
+        cluster.update(str(update_cluster_stop_login_config))
+
+    # wait for LoginNodes gracetime_period
+    for _i in range(5):
+        result = remote_command_executor.run_remote_command(
+            "sudo /opt/parallelcluster/scripts/slurm/check_login_nodes_stopped.sh",
+            raise_on_error=False,
+        )
+        exit_code = result.return_code
+        if exit_code == 0:
+            break
+        else:
+            sleep(180)
```

Comment:
Here we usually use the `@retry` decorators. Please check for examples in the code.

Reply:
Done!